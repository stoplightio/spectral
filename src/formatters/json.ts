import { IRuleResult } from '../types';

export const json = (results: IRuleResult[]): string => {
  const outputJson = results.map(result => {
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
