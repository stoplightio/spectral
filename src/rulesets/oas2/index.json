{
  "extends": ["spectral:oas"],
  "formats": ["oas2"],
  "rules": {
    "api-host": {
      "description": "OpenAPI `host` must be present and non-empty string.",
      "recommended": true,
      "type": "style",
      "given": "$",
      "then": {
        "field": "host",
        "function": "truthy"
      },
      "tags": [
        "api"
      ]
    },
    "api-schemes": {
      "description": "OpenAPI host `schemes` must be present and non-empty array.",
      "recommended": true,
      "type": "style",
      "given": "$",
      "then": {
        "field": "schemes",
        "function": "schema",
        "functionOptions": {
          "schema": {
            "items": {
              "type": "string"
            },
            "minItems": 1,
            "type": "array"
          }
        }
      },
      "tags": [
        "api"
      ]
    },
    "host-not-example": {
      "description": "Server URL should not point at `example.com`.",
      "recommended": false,
      "type": "style",
      "given": "$",
      "then": {
        "field": "host",
        "function": "pattern",
        "functionOptions": {
          "notMatch": "example\\.com"
        }
      }
    },
    "host-trailing-slash": {
      "description": "Server URL should not have a trailing slash.",
      "recommended": true,
      "type": "style",
      "given": "$",
      "then": {
        "field": "host",
        "function": "pattern",
        "functionOptions": {
          "notMatch": "/$"
        }
      }
    },
    "model-description": {
      "description": "Definition `description` must be present and non-empty string.",
      "recommended": false,
      "type": "style",
      "given": "$..definitions[*]",
      "then": {
        "field": "description",
        "function": "truthy"
      }
    },
    "oas2-parameter-description": {
      "description": "Parameter objects should have a `description`.",
      "recommended": false,
      "type": "style",
      "given": "$..[?(@parentProperty === 'parameters' && @.in)]",
      "then": {
        "field": "description",
        "function": "truthy"
      },
      "tags": [
        "parameters"
      ]
    },
    "oas2-operation-security-defined": {
      "description": "Operation `security` values must match a scheme defined in the `securityDefinitions` object.",
      "recommended": true,
      "type": "validation",
      "given": "$",
      "then": {
        "function": "oasOpSecurityDefined",
        "functionOptions": {
          "schemesPath": [
            "securityDefinitions"
          ]
        }
      },
      "tags": [
        "operation"
      ]
    },
    "valid-example-in-parameters": {
      "description": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "recommended": true,
      "severity": 0,
      "type": "validation",
      "given": "$..parameters..[?(@.in == 'body')]..[?(@property !== 'properties' && @.example && ( @.type || @.format || @.$ref ))]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$"
        }
      }
    },
    "valid-example-in-definitions": {
      "description": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "recommended": true,
      "severity": 0,
      "type": "validation",
      "given": "$..definitions..[?(@property !== 'properties' && @.example && (@.type || @.format || @.$ref))]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$"
        }
      }
    },
    "oas2-anyOf": {
      "description": "OpenAPI v3 keyword `anyOf` detected in OpenAPI v2 document.",
      "message": "anyOf is not available in OpenAPI v2, it was added in OpenAPI v3",
      "recommended": true,
      "type": "validation",
      "given": "$..[?(@.anyOf)]",
      "then": {
        "field": "anyOf",
        "function": "undefined"
      },
      "tags": [
        "schema"
      ]
    },
    "oas2-oneOf": {
      "description": "OpenAPI v3 keyword `oneOf` detected in OpenAPI v2 document.",
      "message": "oneOf is not available in OpenAPI v2, it was added in OpenAPI v3",
      "recommended": true,
      "type": "validation",
      "given": "$..[?(@.oneOf)]",
      "then": {
        "field": "oneOf",
        "function": "undefined"
      },
      "tags": [
        "schema"
      ]
    },
    "oas2-schema": {
      "description": "Validate structure of OpenAPI v2 specification.",
      "message": "{{error}}",
      "type": "validation",
      "severity": 0,
      "recommended": true,
      "given": "$",
      "then": {
        "function": "schema",
        "functionOptions": {
          "schema": {
            "$ref": "./schemas/main.json"
          }
        }
      },
      "tags": [
        "schema"
      ]
    }
  }
}
