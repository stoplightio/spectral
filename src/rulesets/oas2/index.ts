import { readRulesFromRulesets } from '../reader';

export { commonOasFunctions as oas2Functions } from '../oas';

export const rules = async () => readRulesFromRulesets(require.resolve('./index.json'));
