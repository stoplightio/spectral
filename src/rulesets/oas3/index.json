{
  "extends": ["spectral:oas"],
  "formats": ["oas3"],
  "rules": {
    "api-servers": {
      "description": "OpenAPI `servers` must be present and non-empty array.",
      "recommended": true,
      "type": "style",
      "given": "$",
      "then": {
        "field": "servers",
        "function": "schema",
        "functionOptions": {
          "schema": {
            "items": {
              "type": "object"
            },
            "minItems": 1,
            "type": "array"
          }
        }
      },
      "tags": [
        "api"
      ]
    },
    "model-description": {
      "description": "Model `description` must be present and non-empty string.",
      "recommended": false,
      "type": "style",
      "given": "$.components.schemas[*]",
      "then": {
        "field": "description",
        "function": "truthy"
      }
    },
    "oas3-operation-security-defined": {
      "description": "Operation `security` values must match a scheme defined in the `components.securitySchemes` object.",
      "recommended": true,
      "type": "validation",
      "given": "$",
      "then": {
        "function": "oasOpSecurityDefined",
        "functionOptions": {
          "schemesPath": [
            "components",
            "securitySchemes"
          ]
        }
      },
      "tags": [
        "operation"
      ]
    },
    "oas3-parameter-description": {
      "description": "Parameter objects should have a `description`.",
      "recommended": false,
      "type": "style",
      "given": "$..[?(@parentProperty !== 'links' && @.parameters)]['parameters'].[?(@.in)]",
      "then": {
        "field": "description",
        "function": "truthy"
      },
      "tags": [
        "parameters"
      ]
    },
    "server-not-example.com": {
      "description": "Server URL should not point at `example.com`.",
      "recommended": false,
      "type": "style",
      "given": "$.servers[*]",
      "then": {
        "field": "url",
        "function": "pattern",
        "functionOptions": {
          "notMatch": "example\\.com"
        }
      }
    },
    "server-trailing-slash": {
      "description": "Server URL should not have a trailing slash.",
      "recommended": true,
      "type": "style",
      "given": "$.servers[*]",
      "then": {
        "field": "url",
        "function": "pattern",
        "functionOptions": {
          "notMatch": "/$"
        }
      }
    },
    "valid-oas-example-in-parameters": {
      "description": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "recommended": true,
      "severity": 0,
      "type": "validation",
      "given": "$..parameters..[?(@.example && @.schema)]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$.schema"
        }
      }
    },
    "valid-oas-example-in-headers": {
      "description": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "recommended": true,
      "severity": 0,
      "type": "validation",
      "given": "$..headers..[?(@.example && @.schema)]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$.schema"
        }
      }
    },
    "valid-oas-example-in-content": {
      "description": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "recommended": true,
      "severity": 0,
      "type": "validation",
      "given": "$..content..[?(@.example && @.schema)]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$.schema"
        }
      }
    },
    "valid-schema-example-in-parameters": {
      "summary": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "recommended": true,
      "severity": 0,
      "type": "validation",
      "given": "$..parameters..[?(@property !== 'properties' && @.example && (@.type || @.format || @.$ref))]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$"
        }
      }
    },
    "valid-schema-example-in-headers": {
      "summary": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "severity": 0,
      "recommended": true,
      "type": "validation",
      "given": "$..headers..[?(@property !== 'properties' && @.example && (@.type || @.format || @.$ref))]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$"
        }
      }
    },
    "valid-example-in-schemas": {
      "summary": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "severity": 0,
      "recommended": true,
      "type": "validation",
      "given": "$.components.schemas..[?(@property !== 'properties' && @.example && (@.type || @.format || @.$ref))]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$"
        }
      }
    },
    "valid-schema-example-in-content": {
      "summary": "Examples must be valid against their defined schema.",
      "message": "{{error}}",
      "severity": 0,
      "recommended": true,
      "type": "validation",
      "given": "$..content..[?(@property !== 'properties' && @.example && (@.type || @.format || @.$ref))]",
      "then": {
        "function": "schemaPath",
        "functionOptions": {
          "field": "example",
          "schemaPath": "$"
        }
      }
    },
    "oas3-schema": {
      "description": "Validate structure of OpenAPI v3 specification.",
      "message": "{{error}}",
      "type": "validation",
      "severity": 0,
      "recommended": true,
      "given": "$",
      "then": {
        "function": "schema",
        "functionOptions": {
          "schema": {
            "$ref": "./schemas/main.json"
          }
        }
      },
      "tags": [
        "schema"
      ]
    }
  }
}
