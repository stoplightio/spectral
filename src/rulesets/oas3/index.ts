import { readRulesFromRulesets } from '../reader';

export { commonOasFunctions as oas3Functions } from '../oas';

export const rules = async () => readRulesFromRulesets(require.resolve('./index.json'));
