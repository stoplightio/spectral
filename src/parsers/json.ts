import { getLocationForJsonPath, JsonParserResult, parseWithPointers, trapAccess } from '@stoplight/json';
import { IParser } from './types';

export const parseJson = (input: string) =>
  parseWithPointers(input, { ignoreDuplicateKeys: false, preserveKeyOrder: true });

export const Json: IParser<JsonParserResult<unknown>> = {
  parse: parseJson,
  getLocationForJsonPath,
  trapAccess,
};
