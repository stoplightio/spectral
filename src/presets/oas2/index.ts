const merge = require('lodash.merge');

import { IPreset } from '../../types';
import { preset as commonOas } from '../oas';
import * as schema from './schemas/main.json';

export const oas2Preset = (): IPreset => {
  return {
    name: 'oas2',
    rules: merge(commonOas().rules, {
      oas2: {
        'oas2-schema': {
          type: 'validation',
          summary: 'Validate structure of OpenAPIv2 specification.',
          enabled: true,
          severity: 'error',
          path: '$',
          function: 'schema',
          input: {
            schema,
          },
        },
      },
    }),
  };
};
