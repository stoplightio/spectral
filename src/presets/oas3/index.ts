const merge = require('lodash.merge');

import { IPreset } from '../../types';
import { preset as commonOas } from '../oas';
import * as schema from './schemas/main.json';

export const oas3Preset = (): IPreset => {
  return {
    name: 'oas3',
    rules: merge(commonOas().rules, {
      oas3: {
        'oas3-schema': {
          type: 'validation',
          summary: 'Validate structure of OpenAPIv3 specification.',
          enabled: true,
          severity: 'error',
          path: '$',
          function: 'schema',
          input: {
            schema,
          },
        },
      },
    }),
  };
};
