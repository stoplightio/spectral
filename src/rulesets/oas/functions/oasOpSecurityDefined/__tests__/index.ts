import { RuleType, Spectral } from '../../../../../index';
import { oas2Functions } from '../../../../oas2/index';
import { rules as oas2Rules } from '../../../../oas2/index.json';
import { oas3Functions } from '../../../../oas3/index';
import { rules as oas3Rules } from '../../../../oas3/index.json';

const oas2Ruleset = { functions: oas2Functions(), rules: oas2Rules };
const oas3Ruleset = { functions: oas3Functions(), rules: oas3Rules };

describe('oasOpSecurityDefined', () => {
  describe('oas2', () => {
    const s = new Spectral();
    s.addFunctions(oas2Ruleset.functions || {});
    s.addRules({
      'operation-security-defined': Object.assign(oas2Ruleset.rules['operation-security-defined'], {
        recommended: true,
        type: RuleType[oas2Ruleset.rules['operation-security-defined'].type],
      }),
    });

    test('validate a correct object (just in body)', async () => {
      const results = await s.run({
        securityDefinitions: {
          apikey: {},
        },
        paths: {
          '/path': {
            get: {
              security: [
                {
                  apikey: [],
                },
              ],
            },
          },
        },
      });
      expect(results.length).toEqual(0);
    });

    test('return errors on invalid object', async () => {
      const results = await s.run({
        securityDefinitions: {},
        paths: {
          '/path': {
            get: {
              security: [
                {
                  apikey: [],
                },
              ],
            },
          },
        },
      });

      expect(results).toMatchSnapshot();
    });
  });

  describe('oas3', () => {
    const s = new Spectral();
    s.addFunctions(oas3Ruleset.functions || {});
    s.addRules({
      'operation-security-defined': Object.assign(oas3Ruleset.rules['operation-security-defined'], {
        recommended: true,
        type: RuleType[oas3Ruleset.rules['operation-security-defined'].type],
      }),
    });

    test('validate a correct object (just in body)', async () => {
      const results = await s.run({
        components: {
          securitySchemes: {
            apikey: {},
          },
        },
        paths: {
          '/path': {
            get: {
              security: [
                {
                  apikey: [],
                },
              ],
            },
          },
        },
      });
      expect(results.length).toEqual(0);
    });

    test('return errors on invalid object', async () => {
      const results = await s.run({
        components: {},
        paths: {
          '/path': {
            get: {
              security: [
                {
                  apikey: [],
                },
              ],
            },
          },
        },
      });

      expect(results).toMatchSnapshot();
    });
  });
});
