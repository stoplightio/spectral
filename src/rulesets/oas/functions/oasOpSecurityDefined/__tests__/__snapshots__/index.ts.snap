// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oasOpSecurityDefined oas2 return errors on invalid object 1`] = `
Array [
  Object {
    "code": "operation-security-defined",
    "message": "Operation \`security\` values must match a scheme defined in the \`securityDefinitions\` object.",
    "path": Array [
      "paths",
      "/path",
      "get",
      "security",
      "0",
    ],
    "range": Object {
      "end": Object {
        "character": 24,
        "line": 7,
      },
      "start": Object {
        "character": 10,
        "line": 6,
      },
    },
    "severity": 1,
    "source": undefined,
    "summary": "Operation \`security\` values must match a scheme defined in the \`securityDefinitions\` object.",
  },
]
`;

exports[`oasOpSecurityDefined oas3 return errors on invalid object 1`] = `
Array [
  Object {
    "code": "operation-security-defined",
    "message": "Operation \`security\` values must match a scheme defined in the \`components.securitySchemes\` object.",
    "path": Array [
      "paths",
      "/path",
      "get",
      "security",
      "0",
    ],
    "range": Object {
      "end": Object {
        "character": 24,
        "line": 7,
      },
      "start": Object {
        "character": 10,
        "line": 6,
      },
    },
    "severity": 1,
    "source": undefined,
    "summary": "Operation \`security\` values must match a scheme defined in the \`components.securitySchemes\` object.",
  },
]
`;
